File: algorithm/src/baekjoon/java/boj/BOJ17298.java
package baekjoon.java.boj;

import java.io.*;
import java.util.Stack;
import java.util.StringTokenizer;

public class BOJ17298 {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        int n = Integer.parseInt(br.readLine());

        StringTokenizer st = new StringTokenizer(br.readLine());
        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
            a[i] = Integer.parseInt(st.nextToken());
        }

        int[] result = new int[n];
        Stack<Integer> stack = new Stack<>();
        stack.push(0);
        for (int i = 1; i < n; i++) {
            while (!stack.isEmpty() && a[stack.peek()] < a[i]) {
                result[stack.pop()] = a[i];
            }
            stack.push(i);
        }

        while (!stack.isEmpty()) {
            result[stack.pop()] = -1;
        }

        StringBuffer sb = new StringBuffer();
        for (int i = 0; i < n; i++) {
            sb.append(result[i] + " ");
        }
        bw.write(sb.toString());
        bw.flush();
    }
}

File: algorithm/src/baekjoon/kotlin/boj/BOJ17298.kt
package baekjoon.kotlin.boj

import java.util.*

fun main() {
    val n = readln().toInt()
    val intArr = readln().split(" ").map { it.toInt() }
    val answer = IntArray(n)
    val stack = Stack<Int>()
    stack.push(0)

    for (i in 1 until n) {
        while (!stack.isEmpty() && intArr[i] > intArr[stack.peek()]) answer[stack.pop()] = intArr[i]
        stack.push(i)
    }

    while (!stack.isEmpty()) {
        answer[stack.pop()] = -1
    }

    val sb = StringBuffer()
    for (i in 0 until n) {
        sb.append(answer[i])
        if (i != n - 1) {
            sb.append(" ")
        }
    }
    println(sb)
}
